proposals/0000-type-vs-constraint.rst

```diff
@@ -0,0 +1,302 @@
+Separating Type from Constraint
+==============================
+
+.. author:: Richard Eisenberg and Simon Peyton Jones
+.. date-accepted::
+.. ticket-url::
+.. implemented::
+.. highlight:: haskell
+.. header:: This proposal is `discussed at this pull request <https://github.com/ghc-proposals/ghc-proposals/pull/518>`_.
+.. sectnum::
+.. contents::
+
+Since version 8.0, GHC has supported a rich structure of inhabited types through the
+new primitive ``TYPE``, allowing
+unlifted types to be treated as first-class citizens, leading to the possibility
+of unlifted datatypes and newtypes, already available.
+
+This proposal extends this treatment to constraints, whose inhabitants
+can usefully be viewed as implicit parameters (typically type class dictionaries).
+Just like how the existence of ``TYPE`` directly opened the door to new innovations
+in datatype and newtype definitions, we expect the similar treatment of constraints
+to allow for strict classes and unlifted and unboxed implicit parameters.
+
+In addition, the new treatment proposed here allows GHC's internal language, Core,
+to more properly tell the difference between types and constraints, eliminating a
+troubling class of bugs that keeps recurring.
+
+Motivation
+----------
+
+1. With this proposal, we immediately allow unlifted and unboxed implicit parameters
+   using the existing ``(?x :: Int#)`` syntax. This allows programs to abstract over
+   implicit parameters without suffering inefficiencies due to boxing.
+
+2. This proposal allows a fix to several type-checker bugs that are otherwise elusive.
+   With this proposal, we can eliminate this entire class of bugs, that we otherwise
+   believe will continue to haunt GHC.  Currently in GHC, `Type` and `Constraint` are
+   *distinct* in the typechecker, but *identical* in Core.  That leads to massive duplication;
+   many, many functions over types have two versions, one for the typechecker and one for Core.
+   It also leads to intellectual complexity, with two subtly-different forms of type equality.
+   Worse, nothing maintains the separation, which leads to those elusive bugs.
```

Comment:
> Under this proposal there would still be at least one discrepancy

You mean that there are some ways in which the source language treats types differently to constraints?  Yes, absolutely. If they were treated identically, we would not need to distinguish them :-).    I'm not sure what you'd like to see here... can you suggest some words that would address your thought?

Reply:
No it was not what I meant at all. Sorry for the misunderstanding.

What I meant is:
- In the current situation: `Constraint /= Type` in the source language, but `Constraint = Type` in Core. This creates ample code duplication and subtle bugs.
- In the proposed situation: `x :: C :: Constraint` is invalid in the source language, but it is valid in the in Core. Does this create (ample?) code duplication and subtle bugs?

Maybe it doesn't. You've thought about it for quite a bit, you've probably come to the conclusion that it doesn't. But it's not obvious (at least not to me :slightly_smiling_face: ). This question, I guess, is part me trying to quench my curiosity, part due diligence.

Reply:
Changing the answer to decidable `Constraint ?= Type` is hugely more invasive than allowing something to be derived in the target language that isn't derivable in the source language. The latter is merely observing that compilation from Haskell to Core is not surjective, which is hardly surprising and to a large extent the the point of elaborating dictionary passing!

` x :: C :: Constraint` being admissible is a footgun insofar that explicit dictionary passing means it is possible to violate coherence, but this is already the case thanks to things like `IncoherentInstances` and, again, a cost of doing business with explicit dictionary passing unless we want to elaborate to "setoid hell" or similar (which would require a much fancier core language).

Overall I see nothing worth worrying about here. In your `Constraint` and `Type` alternative, we have the exact same issue as ` x :: C :: Constraint` is still admissable in Core. In making `Constraint = Type` everywhere we simply have a worse version of that foot-gun.

Neither the status quo nor any proposed alternative does any better with this issue.

Reply:
> Changing the answer to decidable `Constraint ?= Type` is hugely more invasive than allowing something to be derived in the target language that isn't derivable in the source language. The latter is merely observing that compilation from Haskell to Core is not surjective, which is hardly surprising and to a large extent the the point of elaborating dictionary passing!

This depends on implementation details. Please be mindful that I know this part of the GHC rather well. Vague hand-wavy answers are not what I'm after, here.

Reply:
We must be misunderstanding each other because nothing I wrote above was supposed to depend on implementation details beyond the basics of what Haskell and Core are, and the desugaring from the former to the latter.

`x :: C :: Constraint` is not a discrepancy in the sense of Haskell observing something which is *wrong* in Core, because that nothing in Haskell observes `x :: C :: Constraint` cannot be written; that is just a meta-theoretic property. In fact, because implicit parameters are constraints in surface Haskell (though they really shouldn't be), arbitrary  `x :: C :: Constraint` can be "resugared" back to surface Haskell well enough to round trip back to the Core with the same behavior modulo strictness.

The "invasiveness" I was referring to was from the user's perspective (the bugs from such a discrepancy where Core and Haskell conclude different and contradictory answers), not the implementation's. Sorry if that was unclear.

Reply:
Arnaud, I think your question is

> In the proposed situation: x :: C :: Constraint is invalid in the source language, but it is valid in the in Core. Does this create (ample?) code duplication and subtle bugs?

I'm confident that the answer is "no".  It is *already* the case that you can't have a term `(e :: ty :: Constraint)`, nor a binder whose type has that kind.  No change there.  It's easy to guarantee: when we start inferring the type of `e` or or `x` we simply give it a unification variable `alpha :: Type`. 

Nothing in the proposal affects any of this.  Does that reassure you?